lib/request.js

```diff
@@ -0,0 +1,149 @@
+     */
+    static publicRequest ({
+        resource,
+        data
+    } = {}) {
+        const path = `/${API_ROOT}/${resource}?${data}`;
+        return _request({ path: path, method: 'GET' });
+    }
+
+    /**
+     * Make a request to a v1 API endpoint
+     * @param {string} resource API resource to access
+     * @param {string} method HTTP method
+     * @param {string} data URL encoded querystring
+     * @returns {Promise.<object|KlaviyoApiError>} Parsed JSON response from the API | Non-200 response received.
+     */
+    static v1Request({
+        resource,
+        method,
+        data=null
+    } = {}) {
+        const path = `/${API_ROOT}/${V1_API}/${resource}?${data}`;
+        const headers = {
+            'Content-Type': 'application/x-www-form-urlencoded'
+        };
+        return _request({ path: path, method: method, headers: headers });
+    }
+
+    /**
+     * Make a request to a v2 API endpoint.
+     * @param {string} resource API resource to access.
+     * @param {string} method HTTP method.
+     * @param {string} data JSON encoded string.
+     * @returns {Promise.<object|KlaviyoApiError>} Parsed JSON response from the API | Non-200 response received.
+     */
+    static v2Request({
+        resource,
+        method,
+        data
+    } = {}) {
+        const path = `/${API_ROOT}/${V2_API}/${resource}`;
+        const headers = {
+            'Content-Type': 'application/json',
+            'Content-Length': data.length
+        };
+        return _request({ path: path, method: method, body: data, headers: headers });
+    }
+}
+
+/**
+ * Perform an HTTP request
+ * @param {string} path Path to the API resource.
+ * @param {string} method HTTP method.
+ * @param {object} headers HTTP headers.
+ * @param {string} body JSON encoded string.
+ * @returns {Promise.<object|KlaviyoApiError>} Parsed JSON response from the API | Non-200 response received.
+ * @throws {KlaviyoAuthenticationError} Invalid private API token.
+ * @throws {KlaviyoRateLimitError} API rate limit encountered.
+ * @throws {KlaviyoServerError} 50X error from Klaviyo.
+ * @throws {KlaviyoApiError} Other non-200 status code encountered.
+ */
+async function _request({
+    path, 
+    method='GET', 
+    headers={},
+    body=null
+} = {}) {
+
+    if (!headers['User-agent']) { headers['User-agent'] = `Klaviyo-Nodejs/${PACKAGE_VERSION}`; }
+
+    const options = {
+        hostname: KLAVIYO_API_HOSTNAME,
+        port: HTTPS_PORT,
+        path: path,
+        method: method,
+        headers: headers,
+        timeout: 30000 //maybe make this a configurable value?
+    };
+
+    return new Promise((resolve, reject) => {
+        const req = https.request(options, (res) => {
+            let data = [];
+    
+            res.on('data', chunk => {
+                data.push(chunk);
+            });
+
+            res.on('end', () => {
+                data = processBuffer(data);
+                if (res.statusCode == 403) {
```

Comment:
you could def just do a helper method like handleResponseCodes were it would just take the res.statusCode and data as an arg.. it either throw's or nothing and we can continue on in the code